File: golaang/graph/graph.go
package main

import "fmt"



type Edge struct {
	To string
	Weight int
}

type Graph struct {
	adjacency map[string][]Edge
	directed bool
}

func NewGraph(directed bool) *Graph {
	return &Graph{
		adjacency: make(map[string][]Edge),
		directed: directed,
	}
}

func (g *Graph) AddVertex(vertex string) {
	if _, exists := g.adjacency[vertex]; !exists {
		g.adjacency[vertex] = []Edge{}
	}
}

func (g *Graph) AddEdge(u, v string, weight int) {
	g.AddVertex(u)
	g.AddVertex(v)
	g.adjacency[u] = append(g.adjacency[u], Edge{To: v, Weight: weight})
}

func (g *Graph) RemoveEdge(u, v string) {
	g.adjacency[u] = filterEdge(g.adjacency[u], v)
}

func filterEdge(edges []Edge, target string) []Edge  {
	filtered := []Edge{}
	for _, edge := range edges {
		if edge.To != target {
			filtered = append(filtered, edge)
		}
	}

	return filtered
}

func (g *Graph) RemoveVertex(vertex string) {
	delete(g.adjacency, vertex)
	for u, edges := range g.adjacency {
		g.adjacency[u] = filterEdge(edges, vertex)
	}
}

func (g *Graph) PrintGraph() {
	for vertex, edges := range g.adjacency {
		fmt.Printf("%s ->", vertex)

		for _, edge := range edges {
			fmt.Printf("%s(%d)", edge.To, edge.Weight)
		}

		fmt.Println()
	}
}

func main() {
	graph := NewGraph(false) // false = undirected

	graph.AddEdge("A", "B", 2)
	graph.AddEdge("A", "C", 3)
	graph.AddEdge("B", "D", 1)
	graph.AddEdge("C", "D", 4)

	graph.RemoveEdge("B", "D")

	graph.PrintGraph()
}

File: typescript/graph/graph.ts


type Edge = { node: string, weight: number }


class Graph {

    private graph : Map<string, Edge[]> = new Map()
    private directed: Boolean = false

    constructor(directed: false) {
        this.directed = directed
    }

    addVertex(vertex: string) {
        if (!this.graph.has(vertex)) {
            this.graph.set(vertex, [])
        }
    }

    addEdge(u: string, v: string, weight: number=1) {

        this.addVertex(u)
        this.addVertex(v)

        this.graph.get(u)!.push({node: v, weight})

        if (!this.directed) {
            this.graph.get(v)!.push({ node: u, weight })
        }

        // this.graph.set(u, (this.graph.get(u) || []).concat(""))

        // if (!this.directed) {
        //     this.graph.set(v, (this.graph.get(v) || []).concat(""))
        // }
    }
 
    removeEdge(u: string, v: string) {
        this.graph.set(
            u, 
            this.graph.get(u)!.filter(edge => edge.node !== v)
        )

        if (!this.directed) {
            this.graph.set(
                v,
                this.graph.get(v)!.filter(edge => edge.node !== u)
            )
        }
    }

    removeVertex(vertex: string) {
        this.graph.delete(vertex);
        for (const [key, edges] of this.graph) {
            this.graph.set(
                key, 
                edges.filter(edge => edge.node !== vertex)
            )
        }
    }

    printGraph() {
        for (const [vertex, edges] of this.graph.entries()) {
          const edgeStr = edges.map(e => `${e.node}(${e.weight})`).join(", ");
          console.log(`${vertex} -> ${edgeStr}`);
        }
      }
   
}


const g = new Graph(false); // false = undirected

g.addEdge("A", "B", 2);
g.addEdge("A", "C", 3);
g.addEdge("B", "D", 1);
g.addEdge("C", "D", 4);

g.removeEdge("B", "D");

g.printGraph();
